bgpd/bgp_packet.c

```diff
@@ -2028,6 +2071,81 @@ static int bgp_route_refresh_receive(struct peer *peer, bgp_size_t size)
 				   SUBGRP_STATUS_DEFAULT_ORIGINATE);
 	}
 
+	/* Enhanced Route Refresh (rfc7313) */
+	if (subtype & BGP_ROUTE_REFRESH_BORR) {
+		if (peer->t_refresh_stalepath) {
+			zlog_err(
+				"%s: %d/%d, BoRR msg received, whereas BoRR already received",
+				peer->host, afi, safi);
+			return BGP_PACKET_NOOP;
+		}
+
+		/* Set routes to stale */
+		SET_FLAG(peer->sflags, PEER_STATUS_ENHANCED_REFRESH);
+		bgp_set_stale_route(peer, afi, safi);
+
+		if (peer->status == Established) {
+			BGP_TIMER_ON(peer->t_refresh_stalepath,
+				     bgp_refresh_stalepath_timer_expire,
+				     peer->bgp->refresh_stalepath_time);
+		}
+
+		if (bgp_debug_neighbor_events(peer))
+			zlog_debug(
+				"%s: %d/%d, BoRR received, triggering timer for %u seconds",
+				peer->host, afi, safi,
+				peer->bgp->refresh_stalepath_time);
+
+	} else if (subtype & BGP_ROUTE_REFRESH_EORR) {
+		if (!peer->t_refresh_stalepath) {
+			zlog_err(
+				"%s: %d/%d, EoRR received, whereas no BoRR received",
+				peer->host, afi, safi);
+			return BGP_PACKET_NOOP;
+		}
+		BGP_TIMER_OFF(peer->t_refresh_stalepath);
+		if (bgp_debug_neighbor_events(peer))
+			zlog_debug(
+				"%s: %d/%d, EoRR msg received, stopping BoRR timer",
+				peer->host, afi, safi);
+
+		/* Clear stale route entries from peer */
+		bgp_clear_stale_route(peer, afi, safi);
+
+		if (bgp_debug_neighbor_events(peer))
+			zlog_debug(
+				"%s: %d/%d, EoRR msg received, clearing stale routes",
+				peer->host, afi, safi);
+	} else {
+		/* In response to a "normal route refresh request" from the
+		 * peer, the speaker MUST send a BoRR message.
+		 */
+		bgp_route_refresh_send(peer, afi, safi, 0, 0, 0,
```

Comment:
Sending of route refresh capability for respective AFI-SAFI should only be done if that AFI-SAFI is negotiated with this peer.